File: src/engine/models/Sightline.ts
import { Point3D } from './Point3D';

/**
 * 視線ライン（観測者から目標への視線）
 */
export interface Sightline {
  /** 視線の起点 */
  from: Point3D;

  /** 視線の終点 */
  to: Point3D;

  /** 遮られたか */
  obstructed: boolean;

  /** 遮蔽点（遮られた場合） */
  obstructionPoint?: Point3D;

  /** 遮蔽物名 */
  obstructedBy?: string;

  /** 遮蔽点までの距離（メートル） */
  distanceToObstruction?: number;
}

/**
 * Sightlineを生成するヘルパー関数
 */
export function createSightline(
  from: Point3D,
  to: Point3D,
  obstructed: boolean = false,
  obstructionPoint?: Point3D,
  obstructedBy?: string
): Sightline {
  let distanceToObstruction: number | undefined;

  if (obstructed && obstructionPoint) {
    const dx = obstructionPoint.x - from.x;
    const dy = obstructionPoint.y - from.y;
    const dz = obstructionPoint.z - from.z;
    distanceToObstruction = Math.sqrt(dx * dx + dy * dy + dz * dz);
  }

  return {
    from,
    to,
    obstructed,
    obstructionPoint,
    obstructedBy,
    distanceToObstruction,
  };
}

File: tools/revit/Engine/Models/Sightline.cs
namespace NoRegretHome.Revit.Engine.Models
{
    /// <summary>
    /// 視線ライン
    /// </summary>
    public class Sightline
    {
        public Point3D From { get; set; }
        public Point3D To { get; set; }
        public bool Obstructed { get; set; }
        public Point3D? ObstructionPoint { get; set; }
        public string ObstructedBy { get; set; }
        public double? DistanceToObstruction { get; set; }

        public Sightline(Point3D from, Point3D to, bool obstructed = false,
                        Point3D? obstructionPoint = null, string obstructedBy = "")
        {
            From = from;
            To = to;
            Obstructed = obstructed;
            ObstructionPoint = obstructionPoint;
            ObstructedBy = obstructedBy;

            if (obstructed && obstructionPoint.HasValue)
            {
                var dx = obstructionPoint.Value.X - from.X;
                var dy = obstructionPoint.Value.Y - from.Y;
                var dz = obstructionPoint.Value.Z - from.Z;
                DistanceToObstruction = System.Math.Sqrt(dx * dx + dy * dy + dz * dz);
            }
        }
    }
}
